Sources/Helpers/ScreenOrientationHelper.swift

```diff
@@ -0,0 +1,123 @@
+//
+//  ScreenOrientationHelper.swift
+//  OsmAnd Maps
+//
+//  Created by Skalii on 30.10.2023.
+//  Copyright © 2023 OsmAnd. All rights reserved.
+//
+
+import Foundation
+
+@objc(OAScreenOrientationHelper)
+@objcMembers
+class ScreenOrientationHelper : NSObject {
+    
+    private let settings = OAAppSettings.sharedManager()
+    
+    private var cachedUserInterfaceOrientationMask: UIInterfaceOrientationMask = .all
+    private var cachedCurrentInterfaceOrientation: UIInterfaceOrientation = .unknown
+    
+    private static var sharedHelperInstance: ScreenOrientationHelper?
+    static let screenOrientationChangedKey: String = "screenOrientationChangedKey"
+    static let applicationModeChangedKey: String = "applicationModeChangedKey"
+    
+    static var sharedInstance: ScreenOrientationHelper {
+        get {
+            if sharedHelperInstance == nil {
+                sharedHelperInstance = ScreenOrientationHelper()
+            }
+            return sharedHelperInstance!
+        }
+    }
+    
+    override init() {
+        super.init()
+        updateUserInterfaceOrientationMask()
+        updateCurrentInterfaceOrientation()
+        NotificationCenter.default.addObserver(self,
+                                               selector: #selector(onUserOrientationDidChange),
+                                               name: NSNotification.Name(kNotificationSetProfileSetting),
+                                               object: nil)
+        NotificationCenter.default.addObserver(self,
+                                               selector: #selector(onDeviceOrientationDidChange),
+                                               name: UIDevice.orientationDidChangeNotification,
+                                               object: nil)
+        NotificationCenter.default.addObserver(self,
+                                               selector: #selector(onApplicationModeChanged),
+                                               name: NSNotification.Name(Self.applicationModeChangedKey),
+                                               object: nil)
+    }
+    
+    deinit {
+        NotificationCenter.default.removeObserver(self)
+    }
+    
+    @objc private func onApplicationModeChanged() {
+        updateUserInterfaceOrientationMask()
+        onDeviceOrientationDidChange()
+    }
+    
+    @objc private func onUserOrientationDidChange(notification: Notification) {
+        if let obj = notification.object as? OACommonPreference, obj == settings?.mapScreenOrientation {
+            updateUserInterfaceOrientationMask()
+            onDeviceOrientationDidChange()
```

Comment:
onDeviceOrientationDidChange should be called by corresponding notification only